storm-client/src/jvm/org/apache/storm/metrics2/StormMetricRegistry.java

```diff
@@ -35,38 +45,154 @@
-        return registry.counter(metricName);
+        Counter counter = registry.counter(metricName);
+        saveMetricTaskIdMapping(taskId, metricName, counter);
+        return counter;
     }
 
     public Counter counter(String name, String topologyId, String componentId, Integer taskId, Integer workerPort, String streamId) {
         String metricName = metricName(name, topologyId, componentId, streamId, taskId, workerPort);
-        return registry.counter(metricName);
+        Counter counter = registry.counter(metricName);
+        saveMetricTaskIdMapping(taskId, metricName, counter);
+        return counter;
     }
-    
-    public MetricRegistry registry() {
-        return registry;
+
+    public Counter counter(String name, TopologyContext context) {
+        String metricName = metricName(name, context);
+        Counter counter = registry.counter(metricName);
+        saveMetricTaskIdMapping(context.getThisTaskId(), metricName, counter);
+        return counter;
+    }
+
+    public void metricSet(String prefix, MetricSet set, TopologyContext context) {
+        String baseName = metricName(prefix, context);
+        // Instead of registering the metrics as a set, register them individually.
+        // This allows fetching the individual metrics by type (getTaskGauges())
+        // to work as expected.
+        for (Map.Entry<String, Metric> entry : set.getMetrics().entrySet()) {
+            String metricName = baseName + "." + entry.getKey();
+            Metric metric = registry.register(metricName, entry.getValue());
+            saveMetricTaskIdMapping(context.getThisTaskId(), metricName, metric);
+        }
+    }
+
+    public Timer timer(String name, TopologyContext context) {
+        String metricName = metricName(name, context);
+        Timer timer = registry.timer(metricName);
+        saveMetricTaskIdMapping(context.getThisTaskId(), metricName, timer);
+        return timer;
+    }
+
+    public Histogram histogram(String name, TopologyContext context) {
+        String metricName = metricName(name, context);
+        Histogram histogram = registry.histogram(metricName);
+        saveMetricTaskIdMapping(context.getThisTaskId(), metricName, histogram);
+        return histogram;
+    }
+
+    private void saveMetricTaskIdMapping(Integer taskId, String name, Metric metric) {
+        Map<String, Metric> metrics = taskIdMetrics.computeIfAbsent(taskId, (tid) -> new HashMap<>());
+        metrics.put(name, metric);
+    }
+
+    public Map<String, Gauge> getTaskGauges(int taskId) {
+        Map<String, Gauge> taskGaugeMap = new HashMap<>();
+        Collection<Gauge> gauges = registry.getGauges().values();
+        Map<String, Metric> taskMetrics = taskIdMetrics.getOrDefault(taskId, Collections.emptyMap());
+        for (Map.Entry<String, Metric> entry : taskMetrics.entrySet()) {
+            if (gauges.contains(entry.getValue())) {
```

Comment:
instead of going through every metric and check if it's a gauge, I think we probably just need separate maps for different metric types.